atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraReloadableKvsConfig.java

```diff
@@ -39,7 +42,19 @@ public CassandraReloadableKvsConfig(
 
     @Override
     public CassandraServersConfigs.CassandraServersConfig servers() {
-        return config.servers();
+        CassandraServersConfig servers = chooseServers();
+        Preconditions.checkState(
```

Comment:
@jeremyk-91 is it ok to throw at runtime here (and for `concurrentGetRangesThreadPoolSize` and `poolsize`) ?

Reply:
Hmm, yeah it's not great. Looking at this I think the ideal behaviour is to do what we normally do for runtime configs internally:

- on startup, get a runtime config. If it means the state of the world is bad, explode (throw). Otherwise save the last seen good runtime config in a field here.
- thereafter, when this is called, get a runtime config; if it's consistent then update the last seen one, and if it means the state of the world is inconsistent, log a WARN, reject the config and use the previous value of the runtime config.

The main alternatives would involve
- throw (as here): probably not the worst, but I don't know if this is fully safe.
- just return empty: this will break a lot of the assumptions the c* code makes
- @pkoenig10's suggestion of returning a failed type: this is good, though with a lot of the places where this is used I don't know what to do except re-throw, and I don't know if that is safe.

Apologies this has turned out to be a fair bit more complex than initially scoped - if you'd prefer someone on the team writes some of the code once we've got a solution we like, that can be arranged.

Reply:
@jeremyk-91 Thoughts on just using `Refreshable` here? Then we get all this behavior "for free" without having to reimplement the logic to apply updates, save the last good value, etc.

This would also allow us to keep the server size check, because we would construct a new `ImmutableCassandraKeyValueServiceConfig` on each runtime config update.

We would just need to change the interface of `AtlasDbFactory` to accept a `Refreshable` instead of a `Supplier`.
https://github.com/palantir/atlasdb/blob/fd73863df0d8d65eeaa6ed5801a1b6ef8578199c/atlasdb-api/src/main/java/com/palantir/atlasdb/spi/AtlasDbFactory.java#L58

Fortunately, this should be fairly easy because we're already using `Refreshable` as an argument here.
https://github.com/palantir/atlasdb/blob/fd73863df0d8d65eeaa6ed5801a1b6ef8578199c/atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java#L402

Reply:
sourcegraph search shows usages in AtlasDB Proxy and Backup product only, so I'm fine with this change.